Add helm chart schedule-pod-restart

flannoo opened the pull request:
Add new helm chart for scheduling pod restarts

tomkerkhove reviewed (changes requested):
LGTM, but would need some more tweaks such as:
- A CI workflow similar to https://github.com/CoditEU/charts/blob/main/.github/workflows/ci-azure-pipeline-environment.yml
- Release the first version of this Helm chart (https://github.com/CoditEU/charts/blob/main/CONTRIBUTING.md#shipping-a-new-helm-chart-version)

tomkerkhove commented on cronjob-pod-restart/README.md:
````diff
@@ -0,0 +1,46 @@
+
+```console
+helm repo add coditeu https://coditeu.github.io/charts
+helm install coditeu/cronjob-pod-restart
````
```suggestion
helm repo update
helm install coditeu/cronjob-pod-restart
```

krottiers commented on cronjob-pod-restart/README.md:
```diff
@@ -0,0 +1,46 @@
+# Cronjob Pod Restart
+
+A cronjob that executes a scheduled restart of pods by label.
```
Type corn -> cron

flannoo replied:
Fixed in commit https://github.com/CoditEU/charts/pull/20/commits/698fa508826fa53ef79abcacf872086b761c523b

krottiers reviewed (approved):
LGTM - No further comments

flannoo commented on cronjob-pod-restart/templates/_helpers.tpl:
```diff
@@ -0,0 +1,76 @@
+{{/*
+Create the name of the cluster role to use
+*/}}
+{{- define "cronjob-pod-restart.clusterRole" -}}
```
Fixed in commit https://github.com/CoditEU/charts/pull/20/commits/698fa508826fa53ef79abcacf872086b761c523b

flannoo commented on cronjob-pod-restart/templates/cronjob.yaml:
```diff
@@ -0,0 +1,20 @@
+          serviceAccountName: {{ include "cronjob-pod-restart.serviceAccountName" . }}
+          containers:
+          - name: kubectl-container
+            image: bitnami/kubectl:1.17.13
```
Default value will now use the target Kubernetes version. Can be overriden through the values.yaml file
Fixed in commit https://github.com/CoditEU/charts/pull/20/commits/698fa508826fa53ef79abcacf872086b761c523b

flannoo commented on cronjob-pod-restart/README.md:
```diff
@@ -0,0 +1,46 @@
+| `bindings.role.name`         | Name of the role binding                                                 | `get-pods-role-binding`       |
+| `cronjob.name`               | Name of the cronjob that will be created                                 | `cronjob-delete-pods`         |
+| `cronjob.schedule`           | A cronjob expression that defines the schedule when the job should run   | `0 0 * * *`                   |
+| `cronjob.podLabel`           | The label of the pods that need to be restarted through the cronjob      | None                          |
```
Fixed in commit https://github.com/CoditEU/charts/pull/20/commits/698fa508826fa53ef79abcacf872086b761c523b

flannoo commented on cronjob-pod-restart/templates/cronjob.yaml:
```diff
@@ -0,0 +1,20 @@
+apiVersion: batch/v1beta1
+kind: CronJob
+metadata:
+  name: {{ .Values.cronjob.name }}
```
Fixed in commit https://github.com/CoditEU/charts/pull/20/commits/698fa508826fa53ef79abcacf872086b761c523b

flannoo commented on cronjob-pod-restart/README.md:
```diff
@@ -0,0 +1,46 @@
+|:-----------------------------|:-------------------------------------------------------------------------|:------------------------------|
+| `serviceAccount.name`        | Name of the service account to use                                       | `operations-pod-restart`      |
+| `serviceAccount.annotations` | Annotations for the service account                                      | None                          |
+| `roles.cluster.name`         | Name of the role to create                                               | `get-pods-role`               |
```
Fixed in commit https://github.com/CoditEU/charts/pull/20/commits/698fa508826fa53ef79abcacf872086b761c523b

flannoo commented on cronjob-pod-restart/README.md:
```diff
@@ -0,0 +1,46 @@
+| `serviceAccount.name`        | Name of the service account to use                                       | `operations-pod-restart`      |
+| `serviceAccount.annotations` | Annotations for the service account                                      | None                          |
+| `roles.cluster.name`         | Name of the role to create                                               | `get-pods-role`               |
+| `bindings.role.name`         | Name of the role binding                                                 | `get-pods-role-binding`       |
```
Fixed in commit https://github.com/CoditEU/charts/pull/20/commits/698fa508826fa53ef79abcacf872086b761c523b

flannoo commented on cronjob-pod-restart/values.yaml:
```diff
@@ -0,0 +1,22 @@
+
+bindings:
+  role:
+    name: "get-pods-role-binding"
```
Fixed in commit https://github.com/CoditEU/charts/pull/20/commits/698fa508826fa53ef79abcacf872086b761c523b

flannoo commented on cronjob-pod-restart/values.yaml:
```diff
@@ -0,0 +1,22 @@
+    name: "get-pods-role-binding"
+roles:
+  cluster:
+    name: "get-pods-role"
```
Fixed in commit https://github.com/CoditEU/charts/pull/20/commits/698fa508826fa53ef79abcacf872086b761c523b

flannoo commented:
> LGTM, but would need some more tweaks such as:
> 
> * A CI workflow similar to https://github.com/CoditEU/charts/blob/main/.github/workflows/ci-azure-pipeline-environment.yml
> * Release the first version of this Helm chart (https://github.com/CoditEU/charts/blob/main/CONTRIBUTING.md#shipping-a-new-helm-chart-version)

All requested changes have been implemented:
 - CI workflow has been setup.
 - Created helm chart package and updated index so new chart is included in the repo. 
 - Still need to execute the last step to create a github release: "Create a GitHub release for your new Helm chart version by using the following template", not sure how to do this, so will require some help for that.

tomkerkhove commented:
> * Still need to execute the last step to create a github release: "Create a GitHub release for your new Helm chart version by using the following template", not sure how to do this, so will require some help for that.

It's ok - I'll do this when it's merged - Thanks!

tomkerkhove commented on README.md:
```diff
@@ -2,6 +2,7 @@
 Helm chart repo of common apps that we use:
 
 - **Azure Pipeline Environment** - A Helm chart for creating required resources to use Azure Pipelines' Environment
+- **Cronjob Pod Restart** - A Helm chart for automatically restarting pods with labels based on a cron schedule.
```
```suggestion
- **Schedule Pod Restart** - A Helm chart for automatically restarting pods with labels based on a cron schedule.
```

tomkerkhove commented on schedule-pod-restart/README.md:
```diff
@@ -0,0 +1,48 @@
+# Cronjob Pod Restart
```
```suggestion
# Schedule Pod Restart
```

tomkerkhove commented on schedule-pod-restart/README.md:
```diff
@@ -0,0 +1,48 @@
+| `bindings.role.name`         | Name of the role binding                                                 | `manage-pods-role-binding`        |
+| `cronjob.schedule`           | A cronjob expression that defines the schedule when the job should run   | `0 0 * * *`                       |
+| `cronjob.podLabels           | One or more labels on the pods that need to be restarted. Use format `key1=value1,key2=value2`      | None   |
+| `bitnamiContainerImageTag`   | bitnami/kubectl container image tag                                      | Version of the Kubernetes cluster where the chart will be deployed          |
```
Maybe we should go with `image.tag` to align the default approach & not leak that we are using Bitnami?
```suggestion
| `image.tag`   | Tag of container image that contains `kubectl`                                      | Version of the Kubernetes cluster where the chart will be deployed          |
```

flannoo replied:
Fixed in commit https://github.com/CoditEU/charts/pull/20/commits/807d90c57c6f579bacdf4b908af7e96a4a3a2d35. Note that I couldn't use "tag" (probably a reserved keyword), so changed it to "tagname" => image.tagname

tomkerkhove replied:
Thanks!

tomkerkhove commented on schedule-pod-restart/README.md:
```diff
@@ -0,0 +1,48 @@
+| `role.name`                  | Name of the role to create                                               | `manage-pods-role`                |
+| `bindings.role.name`         | Name of the role binding                                                 | `manage-pods-role-binding`        |
+| `cronjob.schedule`           | A cronjob expression that defines the schedule when the job should run   | `0 0 * * *`                       |
+| `cronjob.podLabels           | One or more labels on the pods that need to be restarted. Use format `key1=value1,key2=value2`      | None   |
```
```suggestion
| `cronjob.podLabels`           | One or more labels on the pods that need to be restarted. Use format `key1=value1,key2=value2`      | None   |
```

tomkerkhove commented on schedule-pod-restart/README.md:
````diff
@@ -0,0 +1,48 @@
+To install the chart with the release name `schedule-pod-restart`:
+
+```console
+$ helm install schedule-pod-restart coditeu/schedule-pod-restart
````
Maybe we should give a more representable example where schedule & labels are added?

For example:
```suggestion
$ helm install schedule-pod-restart coditeu/schedule-pod-restart --set "cronjob.schedule=0 1 * * *" --set cronjob.podLabels=app=example
```

Although I haven't tried it so please make sure to test this first 😅

tomkerkhove commented on schedule-pod-restart/README.md:
```diff
@@ -0,0 +1,48 @@
+| `serviceAccount.annotations` | Annotations for the service account                                      | None                              |
+| `role.name`                  | Name of the role to create                                               | `manage-pods-role`                |
+| `bindings.role.name`         | Name of the role binding                                                 | `manage-pods-role-binding`        |
+| `cronjob.schedule`           | A cronjob expression that defines the schedule when the job should run   | `0 0 * * *`                       |
```
Maybe add a link to https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#cron-schedule-syntax?

```suggestion
| `cronjob.schedule`           | A cronjob expression that defines the schedule when the job should run. Learn more about the format [in the official documentation](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#cron-schedule-syntax).   | `0 0 * * *`                       |
```

tomkerkhove commented on schedule-pod-restart/templates/cronjob.yaml:
```diff
@@ -0,0 +1,20 @@
+apiVersion: batch/v1beta1
+kind: CronJob
+metadata:
+  name: {{ include "schedule-pod-restart.name" . }}
```
I would need to check if this is valid, but if it is - Then it would be nice to add our typical labels as well.
```suggestion
  name: {{ include "schedule-pod-restart.name" . }}  
  labels:
    {{- include "schedule-pod-restart.labels" . | nindent 4 }}
```

flannoo replied:
Added in commit https://github.com/CoditEU/charts/pull/20/commits/807d90c57c6f579bacdf4b908af7e96a4a3a2d35 and tested (so syntax is correct)

tomkerkhove commented on schedule-pod-restart/templates/cronjob.yaml:
```diff
@@ -0,0 +1,20 @@
+            command:
+            - /bin/sh
+            - -c
+            - kubectl delete pods -l {{ .Values.cronjob.podLabels }} --namespace={{ .Release.Namespace }}
```
Thinking out loud and not perse for this PR; but can we do this?
```suggestion
            - kubectl get pods -l {{ .Values.cronjob.podLabels }} --namespace={{ .Release.Namespace }}
            - kubectl delete pods -l {{ .Values.cronjob.podLabels }} --namespace={{ .Release.Namespace }}
```

This would list our matching pods first before taking action which would help if pod 2 of 5 fails to understand what was left to delete.

flannoo replied:
It seems that the bitnami container exits after running 1 kubectl statement, so at the moment it's not possible to run multiple kubectl commands (or at least I couldn't make it work this way, maybe there are workarounds for this). We could create a seperate issue for this and add this functionality later on (and leave the single command for now)?

tomkerkhove replied:
No it's ok - This is fine, thank you!

tomkerkhove commented on schedule-pod-restart/values.yaml:
```diff
@@ -0,0 +1,20 @@
+# Default values for azure-pipeline-environment.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
```
Some pieces of the documented configuration is not here, is that on purpose?

https://github.com/CoditEU/charts/pull/20/files?file-filters%5B%5D=.html&file-filters%5B%5D=.md&file-filters%5B%5D=.tgz&file-filters%5B%5D=.tpl&file-filters%5B%5D=.yaml&file-filters%5B%5D=.yml&file-filters%5B%5D=dotfile#diff-034c0482a68bfdaaeb21123826fb3f5d80fffa1373549cf3da886248cc4cc3e4R34

flannoo replied:
Added missing values in commit https://github.com/CoditEU/charts/pull/20/commits/807d90c57c6f579bacdf4b908af7e96a4a3a2d35

tomkerkhove reviewed (commented):
LGTM, just added a few more improvements/alignements - Sorry!

flannoo commented:
> LGTM, just added a few more improvements/alignements - Sorry!

No problem - better to be thorough :)
I implemented the requested changes and updated the helm chart to version 0.1.1
Let me know if any further changes are needed.

tomkerkhove commented:
This looks perfect, thanks! We would just have to move back to 0.1.0 and remove 0.1.1 since 0.1.0 was not shipped yet.

Once that is done, we can merge!

tomkerkhove commented:
Thanks a ton @flannoo!

tomkerkhove commented on schedule-pod-restart/README.md:
````diff
@@ -0,0 +1,48 @@
+```console
+helm repo add coditeu https://coditeu.github.io/charts
+helm repo update
+helm install coditeu/cronjob-pod-restart
````
```suggestion
helm install schedule-pod-restart coditeu/schedule-pod-restart
```
